src/Radicle/Internal/Input.hs

```diff
@@ -0,0 +1,63 @@
+{-# ANN module "HLint: ignore Use <$>" #-}
+
+module Radicle.Internal.Input
+    ( InputT
+    , Completions
+    , noCompletions
+    , wordCompletions
+    , getInputLine
+    , runInputT
+    , outputStrLn
+
+    , MonadException
+    )
+where
+
+-- Haskeline-style wrapper, since it doesn't *quite* provide the right API
+-- (completion functions cannot be dynamically modified).  It will also enable
+-- a nice abstraction barrier for REPL UI features.
+
+import           Protolude
+
+import           Control.Monad.Trans (MonadTrans(..))
+import qualified Data.Text as T
+import           System.Console.Haskeline (MonadException)
+import qualified System.Console.Haskeline as H
+
+newtype InputT m a = InputT { unInputT :: H.InputT (ReaderT (Completions m) m) a }
+    deriving (Functor, Applicative, Monad)
+
+instance MonadTrans InputT where
+    lift = InputT . lift . lift
+
+instance (MonadIO m) => MonadIO (InputT m) where
+    liftIO = lift . liftIO
+
+newtype Completions m = Completions { getCompletions :: H.CompletionFunc (ReaderT (Completions m) m) }
+
+noCompletions :: (Monad m) => Completions m
+noCompletions = Completions H.noCompletion
+
+wordCompletions :: (Monad m) => [Char] -> [Text] -> Completions m
+wordCompletions whitespace words = Completions $ H.completeWord Nothing whitespace go
+    where
+    go s = pure . fmap H.simpleCompletion
+         . filter (s `isPrefixOf`)
+         $ map T.unpack words
+{-# ANN wordCompletions "HLint: ignore Use String" #-}
+
+getInputLine :: (Monad m, MonadException m) => Completions m -> Text -> InputT m (Maybe Text)
+getInputLine compl prompt = InputT $
+    (fmap.fmap) toS $ H.mapInputT (local (const compl)) (H.getInputLine (T.unpack prompt))
```

Comment:
Not actually I disagree with hlint here (`fmap toS <$>` seems perfectly nice). I don't have very strong inclinations wrt hlint, but I'd say generally if you think something is a bad hint, I'd add ignore it in `.hlint.yaml`.

Reply:
This is [SEC](http://conal.net/blog/posts/semantic-editor-combinators) style.  That is, if you have an `f :: A -> B`, and you want to get `Foo (Bar (Baz A)) -> Foo (Bar (Baz B))`, you use `(mapFoo.mapBar.mapBaz) f`.  Most of those `map*`s can be `fmap`, but there are others such as `first` or `left`, which spell out the path to the thing you want to change.  If you still have strong feelings I will give in, I just think this is a nice idiom.